### PR TITLE
[Search v1] Create Search widget

### PR DESCRIPTION
**Pull request by Kicu**
### Details
 - this PR implements this part of the doc: https://docs.google.com/document/d/1u6tBxmoKAEc70jvyZHaKykcfyC5iyEaewfgYXUSSbBc/edit#bookmark=kix.ndabi93vuj44
 - this handles onyx keys + types, data loading logic and empty states logic
 - it only introduces data-related parts of Search component - the actual list with items will come in a separate PR
 - unfortunately there is no api yet, so I had to test it with moc data provided in the doc: https://docs.google.com/document/d/1u6tBxmoKAEc70jvyZHaKykcfyC5iyEaewfgYXUSSbBc/edit#bookmark=id.scrttsoo8ugn 

### Fixed Issues

$ https://github.com/Expensify/App/issues/39879
PROPOSAL:

### Tests

- to enter new Search page go to Settings > Troubleshoot > New Search Page
- by default you should see the "empty" view
- since there is no backend yet, in order to see that onyx correctly displays the data you need to copy-paste Onyx.set command (found below) into browser console - 2 element ids should appear 
- in general I suggest to review code since without data from backend it's hard to fully test how this looks :(
<details>
<summary>fake data code snippet</summary>

```js
// run this code in browser console to insert fake data:
// hash is the hashed param from URL, by default it should be 96673 for string "all"
Onyx.set(`snapshot_${hash}`, {
            search: {
                offset: 0,
                type: 'transaction',
                hasMoreResults: false,
            },
            data: {
                transactions_1234: {
                    receipt: {source: 'http...'},
                    hasEReceipt: false,
                    created: '2024-04-11 00:00:00',
                    amount: 12500,
                    type: 'cash',
                    reportID: '1',
                    transactionThreadReportID: '2',
                    transactionID: '1234',
                },
                transactions_5555: {
                    receipt: {source: 'http...'},
                    hasEReceipt: false,
                    created: '2024-04-11 00:00:00',
                    amount: 12500,
                    type: 'cash', // not present in live data (data outside of snapshot_)
                    reportID: '1',
                    transactionThreadReportID: '2',
                    transactionID: '5555',
                },
            },
        })
```
</details>

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

Screenshot 2024-04-25 at 16 21 50

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

Screenshot 2024-04-25 at 16 00 54

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

### with temp data

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

File: src/components/SelectionList/TemporaryExpenseListItem.tsx
Comment (WojtekBoman) [inline src/components/SelectionList/TemporaryExpenseListItem.tsx:10]: Replace the inline style with styles.p2

File: src/pages/Search/SearchPageBottomTab.tsx
Comment (WojtekBoman) [inline src/pages/Search/SearchPageBottomTab.tsx:21]: Replace the inline style with styles.pt0

Comment (Kicu) [reply src/pages/Search/SearchPageBottomTab.tsx]: will remove for now because this specific padding fix will be added in a separate PR done by Carlos

File: src/components/Search.tsx
Comment (Kicu) [inline src/components/Search.tsx:43]: I understand this looks bad, but Im keeping it here for testing purposes to help reviewers. Unfortunately there is no api for this feature done yet.

Once this gets approve I can push a commit to remove this comment.

**Comment by Kicu** [comment]
Ready to review but I have a TS error which I do not yet understand, I will be fixing it tomorrow

File: src/ONYXKEYS.ts
Comment (luacmartins) [inline src/ONYXKEYS.ts:361]:
```suggestion
        SNAPSHOT: 'snapshot_'
```

**Review by luacmartins — CHANGES_REQUESTED** [review]

File: src/components/Search.tsx
Comment (luacmartins) [inline src/components/Search.tsx:50]: We need to connect to Onyx using the hash, not the query

File: src/components/Search.tsx
Comment (luacmartins) [inline src/components/Search.tsx:42]: Let's add this as a comment on the PR and remove it from the code

File: src/components/Search.tsx
Comment (luacmartins) [inline src/components/Search.tsx:54]: Hmm I found it to be a bit weird that refreshing the page doesn't trigger a call to search, but that's something we could iron out later in a follow up if needed.

Comment (Kicu) [reply src/components/Search.tsx:54]: I believe the `search()` gets called, but its the implementation of `isNetworkOffline` thats breaking it.
I did some logging and in the beginning the NETWORK sets `isOffline` to true for a split second, and that is when the useEffect triggers. Then a bit later `isOffline` is correctly set to `false` but by then effect has already run and no search call was made.

Not sure how to fix this, need some guidance from you. Is that the intended behavior of `ONYXKEYS.NETWORK`?

File: src/libs/SearchUtils.ts
Comment (Kicu) [reply src/libs/SearchUtils.ts:22]: @luacmartins that is actually some TS work since the types are not so trivial to handle here, that is why I omitted it for now to speed things up 😅

**Comment by Kicu** [comment]
@luacmartins thanks for the comments, I implemented almost all the changes you requested.

There is one thing which I would like to do in a separate PR, and that is the proper implementation of `getListItem`/`getSections`. In order to implement them using the mapping object, probably some generic TS types will have to be introduced, because you need to bind them together using an interface. Otherwise if we just return `typesMap[type].ListItem` this is `any`/`unknown` to TS.

However the next issue to do is actually the items list + TransactionLIstItem component, and those will have to be bound via interface, because thats is what `SelectionList` expects anyway.
See: https://github.com/Expensify/App/blob/main/src/components/SelectionList/index.tsx#L8

So I would prefer to keep this PR as is and fix this in another PR, where I would have to do a common interface anyways. It will be easier to do it then, right now it will require a bit of TS-hacking to achieve this

File: src/components/Search.tsx
Comment (luacmartins) [reply src/components/Search.tsx:23]: I don't think that's a blocker for now. Let's move on and address this as a follow up.

**Comment by luacmartins** [comment]
That makes sense, let's address the proper implementation of those methods in a follow up then!

**Comment by luacmartins** [comment]
@dukenv0307 could you please prioritize reimbursing this PR when you're online? Thank you!

**Review by luacmartins — APPROVED** [review]
LGTM and tests well with the API

File: src/pages/Search/SearchFilters.tsx
Comment (luacmartins) [reply src/pages/Search/SearchFilters.tsx:44]: Let's not block on this. We can address this in a follow up @Kicu @WojtekBoman

**Comment by luacmartins** [comment]
Nice! Let's not block on that comment. I can post the screenshots since I have access to the API internally

**Review by dukenv0307 — APPROVED** [review]

**Review by luacmartins — APPROVED** [review]

**Pull request merged**
